R/get_join_keys.R

```diff
@@ -3,57 +3,21 @@
 #' @return Either `JoinKeys` object or `NULL` if no join keys
 #' @export
 get_join_keys <- function(data) {
-  UseMethod("get_join_keys", data)
-}
```

Comment:
I think we still need to support this generic. As `get_join_keys.tdata` is still being used in `teal`. I tried to define this in `teal` but I see CRAN NOTES complaining in other places:
```r
N  checking R code for possible problems (9.8s)
   render_tab_header: no visible global function definition for
     ‘get_join_keys’
   render_tab_table: no visible global function definition for
     ‘get_join_keys’
   srv_a_pca : <anonymous>: no visible global function definition for
     ‘get_join_keys’
   srv_a_regression : <anonymous>: no visible global function definition
     for ‘get_join_keys’
   srv_distribution : <anonymous>: no visible global function definition
     for ‘get_join_keys’
   srv_g_bivariate : <anonymous>: no visible global function definition
     for ‘get_join_keys’
   srv_g_response : <anonymous>: no visible global function definition for
     ‘get_join_keys’
   srv_g_scatterplot : <anonymous>: no visible global function definition
     for ‘get_join_keys’
   srv_g_scatterplotmatrix : <anonymous>: no visible global function
     definition for ‘get_join_keys’
   srv_missing_data : <anonymous>: no visible global function definition
     for ‘get_join_keys’
   srv_outliers : <anonymous>: no visible global function definition for
     ‘get_join_keys’
   srv_t_crosstable : <anonymous>: no visible global function definition
     for ‘get_join_keys’
   srv_tm_g_association : <anonymous>: no visible global function
     definition for ‘get_join_keys’
   Undefined global functions or variables:
     get_join_keys
```

Reply:
Maybe my local has incompatible versions of packages installed, I'll check this once again.

Reply:
Nevermind, This happens in the `main` branch too. So it's not related to this PR

Reply:
https://github.com/insightsengineering/teal.modules.general/actions/runs/6786190730/job/18446154776

Reply:
Interesting case, I guess we need to implement an "inheritted" `join_keys()` on teal

Or support it on `teal.data::join_keys`, but that breaks encapsulation as `teal.data` shouldn't know about `tdata`

Or create a `join_keys()` S3 method that takes `function(x, ...)` and does everything that it does now. (instead of an if-based implementation)


```diff
diff --git a/R/tdata.R b/R/tdata.R
index 5e027ff8e..cf5b7c09e 100644
--- a/R/tdata.R
+++ b/R/tdata.R
@@ -124,29 +124,19 @@ get_code_tdata <- function(data) {
 }
 
 #' Temporary generic method till we support tdata. As `teal.data` deprecated the method
-#' @rdname get_join_keys
-#' @keywords internal
+#' @param ... optional, `JoinKeySet` objects created using the `teal.data::join_key` function.
+#' @return `JoinKeys` object, either created from `JoinKeySet` or the join keys of an existing
+#' object (`teal_data`, `tdata`, ...)
 #' @export
-get_join_keys <- function(data) {
-  UseMethod("get_join_keys", data)
-}
-
-#' Extract `JoinKeys` from `tdata`
-#' @rdname get_join_keys
-#' @param data (`tdata`) object
-#' @keywords internal
-#' @export
-get_join_keys.tdata <- function(data) {
-  attr(data, "join_keys")
-}
+join_keys <- function(...) {
+  x <- rlang::list2(...)
 
+  # Getter
+  if (checkmate::test_list(x, len = 1, types = c("tdata"))) {
+    return(attr(x[[1]], "join_keys"))
+  }
 
-#' @param data object to extract the join keys from
-#' @rdname get_join_keys
-#' @keywords internal
-#' @export
-get_join_keys.default <- function(data) {
-  teal.data::get_join_keys(data)
+  do.call(teal.data::join_keys, x)
 }
 
 #' Function to get metadata from a `tdata` object
```

Reply:
This works (leaving it here just in case)

It actually looks good :smile: and passes all tests

<details>

<summary>Working code for alternative join_keys</summary>

```R
#' Something
#' @export
join_keys <- function(x = NULL, ...) {
  if (is.null(x)) return(new_join_keys())
  UseMethod("join_keys", x)
}

#' @rdname join_keys
#' @export
join_keys.JoinKeys <- function(x, ...) {
  if (missing(...)) return(x)
  join_keys.default(x, ...)
}

#' @rdname join_keys
#' @export
join_keys.teal_data <- function(x, ...) {
  if (missing(...)) return(x@join_keys)
  join_keys.default(x, ...)
}

#' @rdname join_keys
#' @export
join_keys.TealData <- function(x, ...) {
  if (missing(...)) return(x$get_join_keys())
  join_keys.default(x, ...)
}

#' @rdname join_keys
#' @export
join_keys.default <- function(x, ...) {
  x <- append(list(x = x), rlang::list2(...))

  # Constructor
  res <- new_join_keys()
  if (length(x) > 0) {
    join_keys(res) <- x
  }

  res
}
```

</details>